### PR TITLE
feat: add timeZoneName props

### LINKED ISSUE #61: Issue with timezones on iOS
# Bug

In years between 1981-1989 _onChange method return wrong GMT and timezone offset.

The issue is here: https://github.com/react-native-community/react-native-datetimepicker/blob/master/src/datetimepicker.ios.js#L54

1. My timezone is GMT +3. This is important.
2. I've placed console.logs here like that:

```
    if (timestamp) {
      date = new Date(timestamp);
      console.log('_onChange date', date);
      console.log('timezoneoffset', date.getTimezoneOffset());
    }
```
3. Here is what I have 

ezgif com-video-to-gif

So let me describe issue. 

When I pick 1st May 1990 I get 30 Apr and **21:00**. It's ok. Because it is ISO string. In my timezone it is 1 May 00:00. 

But when I change year to 1989 I get 30 Apr and **20:00**. `console.log('timezoneoffset', date.getTimezoneOffset());` still returns -180  which means 3 hours difference. And in this case when I work with date in my timezone, I actually have 30 Apr 23:00. Why that? I see _onChange method handled by native module. This is the only I found.

## Environment info

React native info output:

```bash
System:
    OS: macOS 10.15.1
    CPU: (8) x64 Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz
    Memory: 56.36 MB / 8.00 GB
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 12.12.0 - /usr/local/bin/node
    Yarn: 1.19.1 - /usr/local/bin/yarn
    npm: 6.11.3 - /usr/local/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  SDKs:
    iOS SDK:
      Platforms: iOS 13.2, DriverKit 19.0, macOS 10.15, tvOS 13.2, watchOS 6.1
    Android SDK:
      API Levels: 23, 25, 26, 27, 28
      Build Tools: 27.0.3, 28.0.3
  IDEs:
    Xcode: 11.2.1/11B500 - /usr/bin/xcodebuild
  npmPackages:
    react: ^16.10.1 => 16.12.0 
    react-native: ^0.61.2 => 0.61.5
```

Library version: 2.1.0

### LINKED ISSUE #114: Date picker on iOS has different selected value then Date received in onChange callback
# Bug

DateTimePicker on iOS return in onChange method different then selected one. Different is one day.

## Environment info

React native info output:

```bash
info Fetching system and libraries information...
System:
    OS: macOS 10.15.2
    CPU: (12) x64 Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
    Memory: 128.34 MB / 16.00 GB
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 13.7.0 - /usr/local/bin/node
    Yarn: 1.21.1 - /usr/local/bin/yarn
    npm: 6.13.6 - /usr/local/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  SDKs:
    iOS SDK:
      Platforms: iOS 13.2, DriverKit 19.0, macOS 10.15, tvOS 13.2, watchOS 6.1
  IDEs:
    Android Studio: 3.5 AI-191.8026.42.35.6010548
    Xcode: 11.3.1/11C504 - /usr/bin/xcodebuild
  npmPackages:
    react: ^16.12.0 => 16.12.0 
    react-native: 0.61.4 => 0.61.4 
  npmGlobalPackages:
    create-react-native-app: 2.0.2
    react-native-cli: 2.0.1
```

Library version: 2.2.1

## Steps To Reproduce

1. Create simple app with DateTimePicker.
2. Add item like this:
`<DateTimePicker
                    testID="dateTimePicker"
                    timeZoneOffsetInMinutes={0}
                    value={parseDate(value)}
                    mode={'date'}
                    is24Hour={true}
                    display="default"
                    onChange={(event, date) => {
                        debugger;
                        console.log("Event: " + event.type);
                        console.log("Selected date: " + date));
                />`
3. Run it on iOS simulator.
...

Describe what you expected to happen:

1. Date on picker should be the same as it is in onChange callback function

### LINKED ISSUE #528: timezone offset on android is broken
# Bug report

## Reproducible sample code

use the example app:

## Steps to reproduce

 set tz offset to zero, and then keep opening and closing the calendar picker

the time selected keeps drifting away: 10, 11, 12... and so on

Describe what you expected to happen:

the time should be set, and subsequent opening and closing of the picker should not change the time

## Environment info

`npx react-native info` output:

```bash
System:
    OS: macOS 11.6.1
    CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
    Memory: 180.47 MB / 32.00 GB
    Shell: 5.1.8 - /usr/local/bin/bash
  Binaries:
    Node: 16.8.0 - /usr/local/bin/node
    Yarn: 1.22.11 - /usr/local/bin/yarn
    npm: 7.14.0 - ~/_dev/_own/datetimepicker/node_modules/.bin/npm
    Watchman: 2021.08.23.00 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.11.2 - /usr/local/lib/ruby/gems/2.7.0/bin/pod
  SDKs:
    iOS SDK:
      Platforms: iOS 14.4, DriverKit 20.2, macOS 11.1, tvOS 14.3, watchOS 7.2
    Android SDK:
      API Levels: 23, 25, 26, 27, 28, 29, 30, 31
      Build Tools: 26.0.2, 26.0.3, 28.0.3, 29.0.2, 29.0.3, 30.0.2, 32.0.0
      System Images: android-22 | Google APIs Intel x86 Atom, android-26 | Google APIs Intel x86 Atom, android-26 | Google APIs Intel x86 Atom_64, android-28 | Google Play Intel x86 Atom, android-29 | Intel x86 Atom, android-29 | Intel x86 Atom_64, android-29 | Google APIs Intel x86 Atom, android-30 | Google APIs Intel x86 Atom, android-30 | Google Play Intel x86 Atom, android-31 | Google Play Intel x86 Atom_64
      Android NDK: 19.2.5345600
  IDEs:
    Android Studio: Arctic Fox 2020.3.1 Patch 3 Arctic Fox 2020.3.1 Patch 3
    Xcode: 12.4/12D4e - /usr/bin/xcodebuild
  Languages:
    Java: 1.8.0_292 - /usr/bin/javac
    Python: 2.7.16 - /usr/bin/python
  npmPackages:
    @react-native-community/cli: Not Found
    react: 16.11.0 => 16.11.0
    react-native: ^0.62.2 => 0.62.3
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found```

datetimepicker version: x.x.x

iOS / Android version: x.x

### LINKED ISSUE #547: [Android] timeZoneOffsetInMinutes issue when setting device timezone to behind 1 day
# Bug report

## Summary
When setting device timezone to behind 1 day of existing timezone and setting timeZoneOffsetInMinutes={0} the Android picker is off by one day.

## Reproducible sample code

Using the example project, change the device timezone to Pago GMT -11 (Region: American Samoa).
Set timeZoneOffsetInMinutes={480}

Comment line of code below
 //const sourceMoment = moment.unix(1636797600); line 66
 with
  const sourceMoment = moment();

## Steps to reproduce

1. Click Show Picker

Describe what you expected to happen:

1. 30th Dec should disabled like IOS

IMG_D675CA663B09-1
Screenshot 2021-12-31 at 16 13 31

## Environment info

`npx react-native info` output:

System:
    OS: macOS 11.6.1
    CPU: (8) arm64 Apple M1
    Memory: 102.17 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.8.0 - ~/.nvm/versions/node/v16.8.0/bin/node
    Yarn: Not Found
    npm: 7.14.0 - ~/Downloads/datetimepicker-master/node_modules/.bin/npm
    Watchman: Not Found
  Managers:
    CocoaPods: 1.11.2 - /opt/homebrew/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 21.0.1, iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0
    Android SDK: Not Found
  IDEs:
    Android Studio: 2020.3 AI-203.7717.56.2031.7784292
    Xcode: 13.1/13A1030d - /usr/bin/xcodebuild
  Languages:
    Java: 11.0.12 - /Library/Java/JavaVirtualMachines/jdk-11.0.12.jdk/Contents/Home/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.2 => 17.0.2 
    react-native: ^0.66.4 => 0.66.4 
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found

datetimepicker version: 5.1.0

iOS / Android version: IOS 14.8 / Android 12

### PR DESCRIPTION
**Pull request by wood1986**
# Summary

This PR tries to fix the followings timezone related bugs
1. `timeZoneOffsetInMinutes` does not support daylight saving. 
1. #528 
1. #61

closes #547, closes #61, closes #528, closes #114

My solution to the timezone issues is to let OS interpret time zone by passing a timeZoneName props. JS side should not re-interpret or reconstruct JS Date object based on year, month, day, hour and minutes. Android is doing this and iOS is not. the JS should only need to care UTC timestamp like what iOS is doing.

All the timezone related operation should be handled on native side because OS has the full knowledge of all timezone and library.

The timeZoneName is based on IANA See https://www.iana.org/time-zones in both Android and iOS. But they are not completely identical. It has around 4XX in common.  See https://gist.github.com/mteece/80fff3329074cf90d7991e55f4fc8de4 for NSTimeZone in Objective C and https://garygregory.wordpress.com/2013/06/18/what-are-the-java-timezone-ids/ for TimeZone in Java. 

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ✅     |
| Android |    ✅     |

## Checklist

- [ ] I have tested this on a device and a simulator
- [ ] I added the documentation in `README.md`
- [ ] I updated the typed files (TS and Flow)
- [ ] I added a sample use of the API in the example project (`example/App.js`)
- [ ] I have added automated tests, either in JS or e2e tests, as applicable

### CODE REVIEW

**Comment by wood1986** [comment]
I want to get a feedback from maintainer before writing the test

**Comment by vonovak** [comment]
hello and thanks for the PR!

I'm a little swamped right now with other work that pays the bills and the PR is rather large so it will take me some time to get back to this.

Thank you for you patience! 🙂

**Comment by wood1986** [comment]
Any update?

**Comment by wood1986** [comment]
It has been a month. Any update?

**Review by vonovak — CHANGES_REQUESTED** [review]
Hello and thank you for your PR!
firstly, thank you for taking a stab at this long-term issue. It is one that I was planning on tackling but didn't manage to get to it due to being involved elsewhere. Also, thank you for your patience with getting back to you.

I do agree with the approach of removing JS from the equation and having the date logic live in the native land, so that's good!

I added a few comments that should result into making the code better. 

When they are resolved, docs need to be updated as well. CI also needs to be green before merging.

To wrap up, this is a good direction AFAICT. Please address the review / reply to the comments if you disagree with them. Looking forward to merging! 🚀

File: android/src/main/java/com/reactcommunity/rndatetimepicker/Common.java
Comment (wood1986) [reply android/src/main/java/com/reactcommunity/rndatetimepicker/Common.java:38]: If I do not this, then it will be a O(n) search for each create. Now it does once and use it afterwards in O(log n).

Comment (wood1986) [reply android/src/main/java/com/reactcommunity/rndatetimepicker/Common.java:38]: I want to resolve conversation. Feel free to reactivate it

File: android/src/main/java/com/reactcommunity/rndatetimepicker/RNDatePickerDialogFragment.java
Comment (wood1986) [reply android/src/main/java/com/reactcommunity/rndatetimepicker/RNDatePickerDialogFragment.java:122]: Good call

File: android/src/main/java/com/reactcommunity/rndatetimepicker/TimePickerModule.java
Comment (wood1986) [reply android/src/main/java/com/reactcommunity/rndatetimepicker/TimePickerModule.java:71]: Good call

**Comment by wood1986** [comment]
I fixed the all comments and have not finished yet. Now I can fix the test as you give me a green light and I will do it tomorrow

**Comment by wood1986** [comment]
Hey @vonovak, I am unable to run e2e test both locally and remotely with the same error https://app.circleci.com/pipelines/github/react-native-datetimepicker/datetimepicker/1220/workflows/e7c6c6c6-6cf9-4fa5-8b63-d9cba4df6e6f/jobs/3823/parallel-runs/0/steps/0-115

File: ios/RNDateTimePickerManager.m
Comment (vonovak) [inline ios/RNDateTimePickerManager.m:191]: this warning will currently log even for cases when it shouldn't

File: ios/fabric/RNDateTimePickerComponentView.mm
Comment (vonovak) [inline ios/fabric/RNDateTimePickerComponentView.mm:229]: I believe this will log even when it shouldn't

**Comment by vonovak** [comment]
> Hey @vonovak, I am unable to run e2e test both locally and remotely with the same error 


see https://github.com/react-native-datetimepicker/datetimepicker/pull/758 - that should fix it

**Comment by vonovak** [comment]
@wood1986 if you integrate the latest master, you should be able to run the example app and e2e tests. Let me know if you need any help, thanks! 👍

**Comment by wood1986** [comment]
@vonovak, I am still unable to run the e2e test. It shows

image

**Comment by vonovak** [comment]
There are two ways: 
1) remove the  utilBinaryPaths entry in detoxrc 

2) run the tests on a "vanilla" emulator without Google services installed. You can create one from the command line the same way it's done in circle ci, see the circle ci config file, android/create-avd step.

You'll see the exact command if you look at some previous ci runs in circle ci. I'm away from keyboard so cannot give the exact command.

Thank you! 👍

**Comment by wood1986** [comment]
Hey @vonovak, Now, I can run the test and I have fixed all the tests except tz related test cases. Thanks for that.

I want to completely remove all tzOffsetMinutes as I do not want to spend time on writing test for tzOffsetMinutes. What is your thought?

**Comment by vonovak** [comment]
Please keep the code for tzOffsetMinutes the way it is (don't remove it, don't add tests).

What would be appreciated though is that if someone uses the tzOffsetMinutes prop, they get a depreciation warning, asking them to use the new prop instead. Thank you! 👍

**Comment by wood1986** [comment]
Now, I cannot run e2e test for iOS. It crashes on launch but I can launch it without detox

**Comment by vonovak** [comment]
@wood1986 in CI, the ios tests appear to run just fine, so if they are crashing for you, it seems it's something that is present with you locally.
Is your branch up to date with master branch? It shouldn't really matter but I recently updated detox to latest stable... Not sure really.

Also, not sure why, as the tests are passing on Android, but I see failures with label not matching expectation (ax.id = "utcTime"; text = "2000-01-01T14:44:00Z", expected: toHaveText('2021-11-13T14:44:00Z') and trying to scroll `com.facebook.react.views.scroll.ReactHorizontalScrollView` (such component won't exists on ios).

In case you end up running out of time / patience, let me know and I can take it over the finish line, most of the work is done afaict. Thank you! 👍

**Comment by wood1986** [comment]
Thanks @vonovak 

I have not updated the test for iOS yet. android is completed except one test case in new architecture. 

My branch is up to date and it has all the latest change from master. Please run it on your side and let me know how I can run it locally. I was using iOS 16 and iPhone 14 simulator

**Comment by wood1986** [comment]
Hey @vonovak, Finally, I have just updated all the the tests. The test is a bit unstable. Only you can rerun the test. Please re-review everything as I did a force push.

By the way, I am still unable to run the test on iOS and I was using the build machine result to fix the tests

**Comment by vonovak** [comment]
hello @wood1986 thank you for your work, and sorry to keep you waiting, I'm going to take a look at this next week! 👍

**Review by vonovak — APPROVED** [review]

**Comment by vonovak** [comment]
@wood1986 thank you for your patience and your PR, it'll be released soon! 🙁 💯

**Comment by vonovak** [comment]
:tada: This issue has been resolved in version 7.5.0 :tada:

If this package helps you, consider [sponsoring us](https://github.com/sponsors/react-native-datetimepicker)! :rocket:

**Pull request merged**
